File: .github/scripts/check-markdown-links.py
#!/usr/bin/env python3
from __future__ import annotations

import re
import sys
from pathlib import Path
from urllib.parse import unquote, urlsplit

INLINE_LINK = re.compile(r"!?\[[^\]]*\]\(([^)]+)\)")
REFERENCE_LINK = re.compile(r"^\s*\[[^\]]+\]:\s*(\S+)")
INLINE_CODE = re.compile(r"`[^`]*`")
FENCE = re.compile(r"^\s*(```+|~~~+)")
EXTERNAL_SCHEMES = {"http", "https", "mailto", "tel", "data", "javascript"}


def destination(raw: str) -> str:
    raw = raw.strip()
    if raw.startswith("<") and ">" in raw:
        return raw[1 : raw.index(">")]
    return raw.split(maxsplit=1)[0]


def markdown_targets(text: str):
    in_fence = False
    fence_char = ""
    for line_no, line in enumerate(text.splitlines(), 1):
        match = FENCE.match(line)
        if match:
            marker = match.group(1)
            char = marker[0]
            if not in_fence:
                in_fence = True
                fence_char = char
            elif char == fence_char:
                in_fence = False
                fence_char = ""
            continue
        if in_fence:
            continue

        cleaned = INLINE_CODE.sub("", line)
        for match in INLINE_LINK.finditer(cleaned):
            yield line_no, destination(match.group(1))
        ref = REFERENCE_LINK.match(cleaned)
        if ref:
            yield line_no, destination(ref.group(1))


def local_candidate(repo: Path, source: Path, raw_target: str) -> Path | None:
    if not raw_target or raw_target.startswith("#"):
        return None

    parsed = urlsplit(raw_target)
    if parsed.scheme.lower() in EXTERNAL_SCHEMES or parsed.netloc:
        return None

    path_text = unquote(parsed.path)
    if not path_text:
        return None

    if path_text.startswith("/"):
        relative = Path(path_text.lstrip("/"))
        if not relative.parts:
            return None
        # Root-style web routes (for example /api/health) are not repository links.
        if not (repo / relative.parts[0]).exists() and relative.suffix.lower() != ".md":
            return None
        candidate = repo / relative
    else:
        candidate = source.parent / path_text

    try:
        resolved = candidate.resolve(strict=False)
        resolved.relative_to(repo)
    except (OSError, ValueError):
        return candidate
    return resolved


def main() -> int:
    repo = Path(sys.argv[1] if len(sys.argv) > 1 else ".").resolve()
    failures: list[str] = []

    for source in sorted(repo.rglob("*.md")):
        if ".git" in source.parts:
            continue
        try:
            text = source.read_text(encoding="utf-8")
        except UnicodeDecodeError:
            continue
        for line_no, target in markdown_targets(text):
            candidate = local_candidate(repo, source, target)
            if candidate is None:
                continue
            if not candidate.exists():
                try:
                    display = candidate.relative_to(repo)
                except ValueError:
                    display = candidate
                failures.append(
                    f"{source.relative_to(repo)}:{line_no}: broken local link {target!r} -> {display}"
                )

    if failures:
        print("Broken repository-local Markdown links found:", file=sys.stderr)
        for failure in failures:
            print(f"  {failure}", file=sys.stderr)
        return 1

    print("Repository-local Markdown links are valid.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: .github/scripts/check-markdown-links.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const INLINE_LINK = /!?\[[^\]]*\]\(([^)]+)\)/g;
const REFERENCE_LINK = /^\s*\[[^\]]+\]:\s*(\S+)/;
const INLINE_CODE = /`[^`]*`/g;
const FENCE = /^\s*(```+|~~~+)/;
const EXTERNAL_SCHEMES = new Set(['http', 'https', 'mailto', 'tel', 'data', 'javascript']);

export function destination(raw) {
  const value = raw.trim();
  if (value.startsWith('<') && value.includes('>')) return value.slice(1, value.indexOf('>'));
  return value.split(/\s+/, 1)[0];
}

export function markdownTargets(text) {
  const result = [];
  let inFence = false;
  let fenceChar = '';
  for (const [index, line] of text.split(/\r?\n/).entries()) {
    const fence = line.match(FENCE);
    if (fence) {
      const char = fence[1][0];
      if (!inFence) { inFence = true; fenceChar = char; }
      else if (char === fenceChar) { inFence = false; fenceChar = ''; }
      continue;
    }
    if (inFence) continue;
    const cleaned = line.replace(INLINE_CODE, '');
    for (const match of cleaned.matchAll(INLINE_LINK)) result.push([index + 1, destination(match[1])]);
    const reference = cleaned.match(REFERENCE_LINK);
    if (reference) result.push([index + 1, destination(reference[1])]);
  }
  return result;
}

export function localCandidate(repo, source, rawTarget) {
  if (!rawTarget || rawTarget.startsWith('#')) return null;
  const schemeMatch = rawTarget.match(/^([A-Za-z][A-Za-z0-9+.-]*):(.*)$/s);
  const scheme = schemeMatch?.[1].toLowerCase() ?? '';
  if (EXTERNAL_SCHEMES.has(scheme)) return null;
  if ((!scheme && rawTarget.startsWith('//')) || (scheme && schemeMatch[2].startsWith('//') && !schemeMatch[2].startsWith('///'))) return null;
  let pathText = schemeMatch ? schemeMatch[2] : rawTarget;
  const suffix = pathText.search(/[?#]/);
  if (suffix >= 0) pathText = pathText.slice(0, suffix);
  try { pathText = decodeURIComponent(pathText); } catch { /* retain malformed escapes, as urlsplit does */ }
  if (!pathText) return null;
  let candidate;
  if (pathText.startsWith('/')) {
    const relative = pathText.replace(/^\/+/, '');
    if (!relative) return null;
    const first = relative.split(/[\\/]/)[0];
    if (!fs.existsSync(path.join(repo, first)) && path.extname(relative).toLowerCase() !== '.md') return null;
    candidate = path.join(repo, relative);
  } else candidate = path.resolve(path.dirname(source), pathText.replace(/^\/+/, ''));
  const resolved = path.resolve(candidate);
  const rel = path.relative(repo, resolved);
  if (rel === '..' || rel.startsWith(`..${path.sep}`) || path.isAbsolute(rel)) return resolved;
  return resolved;
}

function walkMarkdown(root) {
  const files = [];
  const walk = (dir) => {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true }).sort((a, b) => a.name < b.name ? -1 : a.name > b.name ? 1 : 0)) {
      if (entry.name === '.git') continue;
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) walk(full);
      else if (entry.isFile() && entry.name.endsWith('.md')) files.push(full);
    }
  };
  walk(root);
  return files.sort((a, b) => {
    const left = path.relative(root, a).split(path.sep).join('/');
    const right = path.relative(root, b).split(path.sep).join('/');
    return left < right ? -1 : left > right ? 1 : 0;
  });
}

function quotedTarget(value) {
  const quote = value.includes("'") && !value.includes('"') ? '"' : "'";
  return `${quote}${value.replaceAll('\\', '\\\\').replaceAll(quote, `\\${quote}`).replaceAll('\n', '\\n').replaceAll('\r', '\\r').replaceAll('\t', '\\t')}${quote}`;
}

function readUtf8(file) {
  try { return new TextDecoder('utf-8', { fatal: true }).decode(fs.readFileSync(file)); }
  catch { return null; }
}

export function check(repo) {
  const failures = [];
  for (const source of walkMarkdown(repo)) {
    const text = readUtf8(source);
    if (text === null) continue;
    for (const [line, target] of markdownTargets(text)) {
      const candidate = localCandidate(repo, source, target);
      if (candidate !== null && !fs.existsSync(candidate)) {
        const display = path.relative(repo, candidate) || candidate;
        failures.push(`${path.relative(repo, source).split(path.sep).join('/')}:${line}: broken local link ${quotedTarget(target)} -> ${display.split(path.sep).join('/')}`);
      }
    }
  }
  return failures;
}

function main() {
  const repo = path.resolve(process.argv[2] ?? '.');
  const failures = check(repo);
  if (failures.length) {
    console.error('Broken repository-local Markdown links found:');
    for (const failure of failures) console.error(`  ${failure}`);
    return 1;
  }
  console.log('Repository-local Markdown links are valid.');
  return 0;
}

if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) process.exitCode = main();
